### PR TITLE
Added client metrics for the CRT HTTP SPI implementation.

### PR DESCRIPTION
**Pull request by JonathanHenson**
## Motivation and Context
Adds metrics for connection managers for the common runtime's implementation of the http SPI.

## Modifications
Requires an upstream update from CRT. I bumped the version in maven to the next version so the CI will fail until the upstream artifacts propagate (that's intentional).

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)

## Checklist

- [ x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x ] Local run of `mvn install` succeeds
- [x ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [x ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java
Comment (zoewangg) [inline http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java:259]: Should we not call report metrics if `asyncRequest.metricCollector()` is not present or it's an instance of NoOpMetricCollector (the default SDK metric collector) to avoid potential performance degradation? We skip metrics collection in netty and apache http client fwiw

https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/NettyRequestMetrics.java#L46-L49

Comment (JonathanHenson) [reply http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java:259]: I was thinking noop metric collector was pretty cheap due to it being a no-op? What's the performance degradation?

Comment (JonathanHenson) [reply http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java:259]: yes looking at NoOpMetricCollector.java, it's just a function call. It seems to me that less branching is a better trade-off than avoiding a java function call that the compiler probably optimizes out anyways. From a performance perspective, an instruction cache miss due to branching is likely to have worse performance profiles than an empty function call. Will defer to your judgement however,

Comment (zoewangg) [reply http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java:259]: Yeah, @millems did benchmark before: https://github.com/aws/aws-sdk-java-v2/pull/1960#issue-664833683 Disabling is slightly better

Comment (JonathanHenson) [reply http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java:259]: i'll happily remove for the sake of consistency with the rest of the code base

Comment (JonathanHenson) [reply http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/AwsCrtAsyncHttpClient.java:259]: plus it would save a JNI call, which is certainly non-trivial in this case

**Review by zoewangg — COMMENTED** [review]
LGTM, we will merge it once crt 0.19.5 is out.

File: pom.xml
Comment (JonathanHenson) [reply pom.xml:116]: when I tag HEAD of the repo, it will kick off CD and will be available as soon as it hits maven.

**Review by zoewangg — APPROVED** [review]

**Comment by TingDaoK** [comment]
The new version of aws-crt-java is now available from maven.

File: http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/CrtResponseAdapter.java
Comment (zoewangg) [inline http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/CrtResponseAdapter.java:132]: Can we add a test to verify the behavior?

https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/aws-crt-client/src/test/java/software/amazon/awssdk/http/crt/AwsCrtHttpClientWireMockTest.java#L46

We can probably stub errors with something like this

https://github.com/aws/aws-sdk-java-v2/blob/639e57ccc612ea40e95cc8556943fe1fc333d39f/http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClientSpiVerificationTest.java#L72-L92

**Review by zoewangg — COMMENTED** [review]
Looks like checkstyle failed
```
[ERROR] /codebuild/output/src262660489/src/github.com/aws/aws-sdk-java-v2/http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/CrtRequestContext.java:22:8: Unused import - software.amazon.awssdk.metrics.NoOpMetricCollector. [UnusedImports]
[ERROR] /codebuild/output/src262660489/src/github.com/aws/aws-sdk-java-v2/http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/CrtRequestContext.java:54:72: '}' at column 72 should be alone on a line. [RightCurlyAlone]
[ERROR] /codebuild/output/src262660489/src/github.com/aws/aws-sdk-java-v2/http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/CrtRequestExecutor.java:39:1: Wrong lexicographical order for 'software.amazon.awssdk.http.crt.internal.request.CrtRequestAdapter' import. Should be before 'software.amazon.awssdk.metrics.NoOpMetricCollector'. [CustomImportOrder]
[ERROR] /codebuild/output/src262660489/src/github.com/aws/aws-sdk-java-v2/http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/CrtRequestExecutor.java:40:1: Wrong lexicographical order for 'software.amazon.awssdk.http.crt.internal.response.CrtResponseAdapter' import. Should be before 'software.amazon.awssdk.metrics.NoOpMetricCollector'. [CustomImportOrder]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:check (checkstyle) on project aws-crt-client: Failed during checkstyle execution: There are 4 errors reported by Checkstyle 8.42 with software/amazon/awssdk/checkstyle.xml ruleset. -> [Help 1]

````

**Comment by zoewangg** [comment]
Merging to mainline PR: https://github.com/aws/aws-sdk-java-v2/pull/3531

**Pull request merged**
